imap/caldav_db.c

```diff
@@ -1468,3 +1468,84 @@ EXPORTED const char *caldav_comp_type_as_string(unsigned comp_type)
     }
 }
 
+static icaltimezone *_get_calendar_tz(const char *mboxname, const char *userid)
+{
+    struct buf attrib = BUF_INITIALIZER;
+    icaltimezone *tz = NULL;
+
+    /*  Check for CALDAV:calendar-timezone-id */
+    const char *prop_annot =
+        DAV_ANNOT_NS "<" XML_NS_CALDAV ">calendar-timezone-id";
+
+    int r = annotatemore_lookupmask(mboxname, prop_annot, userid, &attrib);
+    if (!r && buf_len(&attrib)) {
+        tz = icaltimezone_get_builtin_timezone(buf_cstring(&attrib));
+        buf_free(&attrib);
+        if (tz) return icaltimezone_copy(tz);
+    }
+
+    /*  Check for CALDAV:calendar-timezone */
+    prop_annot = DAV_ANNOT_NS "<" XML_NS_CALDAV ">calendar-timezone";
+
+    r = annotatemore_lookupmask(mboxname, prop_annot, userid, &attrib);
+    if (!r && buf_len(&attrib)) {
+        icalcomponent *ical, *vtz;
+
+        ical = icalparser_parse_string(buf_cstring(&attrib));
+        vtz = icalcomponent_get_first_component(ical, ICAL_VTIMEZONE_COMPONENT);
+        icalcomponent_remove_component(ical, vtz);
+        icalcomponent_free(ical);
+        buf_free(&attrib);
+
+        tz = icaltimezone_new();
+        icaltimezone_set_component(tz, vtz);
+        return tz;
+    }
+
+    return NULL;
+}
+
+EXPORTED icaltimezone *caldav_get_calendar_tz(const char *mboxname,
+                                              const char *userid)
+{
+    icaltimezone *tz = _get_calendar_tz(mboxname, userid);
+
+    if (!tz) {
+        /* Try principal (calendar-home-set) */
+        char *homeset = caldav_mboxname(userid, NULL);
+        tz = _get_calendar_tz(homeset, userid);
+        free(homeset);
+
+        if (!tz) {
+            /* Default to UTC */
+            tz = icaltimezone_copy(icaltimezone_get_utc_timezone());
+        }
+    }
+
+    return tz;
+}
+
+#define CMD_SELFLOATING CMD_READFIELDS \
+    " WHERE mailbox = :mailbox AND dtstart NOT LIKE '%T%' ORDER BY imap_uid;"
+
+EXPORTED int caldav_get_floating(struct caldav_db *caldavdb,
```

Comment:
I think this needs a more complete name - I initially thought it was getting the floating timezone, rather than fetching all the UIDs of floating events.

Reply:
Done